src/generate_explicit_instantiations.py

```diff
@@ -83,121 +87,30 @@ def gen_inst(template, filename, ifdef=None, folder=output_folder, exceptions=No
-gen_inst("class MetisReorder<$id_type, $nnz_type, $value_type>", "preprocess.inc", ifdef="USE_METIS")
-gen_inst("class AMDReorder<$id_type, $nnz_type, $value_type>", "preprocess.inc", ifdef="USE_AMD_ORDER")
-gen_inst("class ReorderHeatmap<$id_type, $nnz_type, $value_type, $float_type>", "preprocess.inc")
-gen_inst("class PermuteOrderTwo<$id_type, $nnz_type, $value_type>", "preprocess.inc")
-gen_inst("class PermuteOrderOne<$id_type, $value_type>", "preprocess.inc", exceptions={"$value_type": ['void']})
-gen_inst("class TransformPreprocessType<format::FormatOrderTwo<$id_type, $nnz_type, $value_type>, format::FormatOrderTwo<$id_type, $nnz_type, $value_type>>", "preprocess.inc")
-gen_inst("class TransformPreprocessType<format::FormatOrderOne<$value_type>, format::FormatOrderOne<$value_type>>", "preprocess.inc")
-gen_inst("class TransformPreprocessType<format::FormatOrderOne<$id_type>, format::FormatOrderOne<$id_type>>", "preprocess.inc")
-gen_inst("class TransformPreprocessType<format::FormatOrderOne<$nnz_type>, format::FormatOrderOne<$nnz_type>>", "preprocess.inc")
-gen_inst("class DegreeDistribution<$id_type, $nnz_type, $value_type, $float_type>", "preprocess.inc")
-gen_inst("class Degrees_DegreeDistribution<$id_type, $nnz_type, $value_type, $float_type>", "preprocess.inc")
-gen_inst("class JaccardWeights<$id_type, $nnz_type, $value_type, $float_type>", "preprocess.inc")
-gen_inst("class MetisPartition<$id_type, $nnz_type, $value_type>", "preprocess.inc", ifdef="USE_METIS")
-gen_inst("class PulpPartition<$id_type, $nnz_type, $value_type>", "preprocess.inc", ifdef="USE_PULP")
-gen_inst("class PatohPartition<$id_type, $nnz_type, $value_type>", "preprocess.inc", ifdef="USE_PATOH")
-gen_inst("class FunctionMatcherMixin<format::Format*>", "preprocess.inc")
-gen_inst("class FunctionMatcherMixin<$id_type*>", "preprocess.inc")
-gen_inst("class FunctionMatcherMixin<$id_type>", "preprocess.inc")
-gen_inst("class FunctionMatcherMixin<$id_type*, ExtractableType>", "preprocess.inc")
-gen_inst("class FunctionMatcherMixin<$id_type, ExtractableType>", "preprocess.inc")
-gen_inst("class GenericPreprocessType<$id_type>", "preprocess.inc")
-gen_inst("class ReorderPreprocessType<$id_type>", "preprocess.inc")
-gen_inst("class PartitionPreprocessType<$id_type>", "preprocess.inc")
-
-# Preprocess (CUDA)
-reset_file("preprocess.inc", cuda_output_folder)
-gen_inst("format::cuda::CUDAArray<$float_type>* "
-         + "RunJaccardKernel<$id_type, $nnz_type, $value_type, $float_type>(format::cuda::CUDACSR<$id_type, $nnz_type, $value_type>*)",
-         "preprocess.inc",
-         ifdef="USE_CUDA",
-         folder=cuda_output_folder)
+#reading class registration file
+with open(class_list, 'r') as read_file:
+    data = json.load(read_file)
+    
+current_filename = ""
+class_instantiation_list = data.get("classes")
+for each_class in class_instantiation_list:
+    #print(each_class)
+    #setting current filename
+    if current_filename != each_class.get("filename"):
```

Comment:
I think this mechanism requires the writers into `class_instantiation_list.json` to make sure they list out all the classes of a file sequentially, which is a reasonable thing to ask, but can still lead to weird errors if a mistake is made. I think a better approach is to store a dictionary of opened files and only reset a file the first time it is encountered.